File: sdk/java/JossPlugin.java
package joss.sdk;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.util.HashMap;
import java.util.Map;
import java.util.function.Function;

/**
 * Robust Java SDK for Joss JP v2 sidecars (joss-rpc-v1).
 * Compatible with standard JVM and GraalVM native-image.
 */
public final class JossPlugin {
    public static final String PROTOCOL = "joss-rpc-v1";

    private final Map<String, Function<String, String>> methods = new HashMap<>();

    public JossPlugin registerMethod(String name, Function<String, String> handler) {
        methods.put(name, handler);
        return this;
    }

    public static void run(Function<String, String> dispatch) throws IOException {
        String request = new String(System.in.readAllBytes(), StandardCharsets.UTF_8);
        String response;
        try {
            response = dispatch.apply(request);
            if (response == null || response.isBlank()) {
                response = "{\"id\":\"\",\"error\":{\"code\":\"EMPTY_RESPONSE\",\"message\":\"Handler returned empty response\"}}";
            }
        } catch (Exception e) {
            response = String.format(
                "{\"id\":\"\",\"error\":{\"code\":\"%s\",\"message\":\"%s\"}}",
                e.getClass().getSimpleName(),
                escapeJson(e.getMessage() != null ? e.getMessage() : e.toString())
            );
        }
        System.out.write(response.getBytes(StandardCharsets.UTF_8));
        System.out.write('\n');
        System.out.flush();
    }

    private static String escapeJson(String raw) {
        if (raw == null) return "";
        return raw.replace("\\", "\\\\").replace("\"", "\\\"").replace("\n", "\\n").replace("\r", "\\r");
    }
}

File: sdk/kotlin/JossPlugin.kt
package joss.sdk

import java.nio.charset.StandardCharsets

/**
 * Idiomatic Kotlin DSL Framework for Joss JP v2 sidecars (joss-rpc-v1).
 * Supports Kotlin/JVM and Kotlin/Native for single-binary plugins.
 */
class JossPlugin private constructor() {
    companion object {
        const val PROTOCOL = "joss-rpc-v1"

        fun define(block: JossPluginBuilder.() -> Unit): JossPluginRunner {
            val builder = JossPluginBuilder()
            builder.block()
            return builder.build()
        }

        fun run(dispatch: (String) -> String) {
            try {
                val rawInput = System.`in`.readBytes().toString(StandardCharsets.UTF_8)
                val response = dispatch(rawInput)
                println(response)
            } catch (e: Throwable) {
                val errorJson = """{"id":"","error":{"code":"${e.javaClass.simpleName}","message":"${escapeJson(e.message)}"}}"""
                println(errorJson)
            }
        }

        private fun escapeJson(raw: String?): String {
            if (raw == null) return ""
            return raw.replace("\\", "\\\\").replace("\"", "\\\"").replace("\n", "\\n").replace("\r", "\\r")
        }
    }
}

class JossPluginBuilder {
    private val handlers = mutableMapOf<String, (String) -> String>()

    fun method(name: String, handler: (String) -> String) {
        handlers[name] = handler
    }

    fun build(): JossPluginRunner {
        return JossPluginRunner(handlers)
    }
}

class JossPluginRunner(private val handlers: Map<String, (String) -> String>) {
    fun run() {
        JossPlugin.run { rawRequest ->
            // Execute method or return error
            handlers.values.firstOrNull()?.invoke(rawRequest) ?: """{"id":"","error":{"code":"UNKNOWN_METHOD","message":"No handler registered"}}"""
        }
    }
}
